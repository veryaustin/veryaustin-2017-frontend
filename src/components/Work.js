import React from 'react';

const Work = () => {
  return (
    <div id="work">This is the Work Component</div>
  );
};

export default Work;
