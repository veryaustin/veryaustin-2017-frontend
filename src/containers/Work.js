import React, { Component, PropTypes } from "react";
import { connect } from "react-redux";
import { bindActionCreators } from "redux";
import { fetchPage } from "../actions/fetchActions";
import Banner from "../components/Banner";
import Section from "../components/Section";
import Loading from "../components/Loading";

class Work extends Component {
  componentDidMount() {
    this.props.fetchPage(this.props.route.path);
  }

  // Helper function to render the various sections
  renderProjects(projectData, index) {
    const {
      title,
      summary,
      description,
      site_url,
      repo_url,
      image_url,
      tools
    } = projectData;

    if (index === 0 || index % 2 === 0) {
      return (
        <Section
          key={index}
          title={title}
          caption={summary}
          description={description}
          image_url={image_url}
          primary_site_url={site_url}
          secondary_site_url={repo_url}
          items={tools}
          dark
        />
      );
    } else {
      return (
        <Section
          key={index}
          title={title}
          caption={summary}
          description={description}
          image_url={image_url}
          primary_site_url={site_url}
          secondary_site_url={repo_url}
          items={tools}
        />
      );
    }
  }

  // Map over the returned data from the action creator and format it on the page
  render() {
    // Destructure the work object & loading/error props
    const {
      work: { title, caption, banner_url, projects = [] },
      isLoading,
      hasErrored
    } = this.props;
    return (
      <div>
        {isLoading ? (
          <Loading />
        ) : (
          <div id="work">
            <Banner
              title={title}
              featured_image={banner_url}
              caption={caption}
            />
            {projects.map(this.renderProjects)}
          </div>
        )}
      </div>
    );
  }
}

// Map State To Props
const mapStateToProps = ({
  pages: { work },
  fetching: { isLoading, error }
}) => {
  return {
    work,
    isLoading,
    hasErrored: error
  };
};

// Bind the actions and dispatch them
function mapDispatchToProps(dispatch) {
  return bindActionCreators({ fetchPage }, dispatch);
}

// Work Props Validation
Work.propTypes = {
  fetchPage: PropTypes.func,
  isLoading: PropTypes.bool,
  hasErrored: PropTypes.string,
  work: React.PropTypes.shape({
    title: PropTypes.string,
    banner_url: PropTypes.string,
    caption: PropTypes.string,
    projects: PropTypes.array
  }),
  projects: PropTypes.shape({
    title: PropTypes.string,
    summary: PropTypes.string,
    description: PropTypes.string,
    primary_url: PropTypes.string,
    secondary_url: PropTypes.string,
    image_url: PropTypes.string,
    tools: PropTypes.array
  }),
  route: PropTypes.object
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Work);
