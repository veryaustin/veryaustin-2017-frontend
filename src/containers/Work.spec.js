import React from "react";
import { shallow } from "enzyme";
import toJson from "enzyme-to-json";
import configureStore from "redux-mock-store";
import Work from "./Work";

describe("Work Component", () => {
  let middlewares, mockStore, store;
  beforeEach(() => {
    middlewares = [];
    mockStore = configureStore(middlewares);
  });

  it("renders the work component", () => {
    store = mockStore({
      pages: {
        work: {}
      },
      fetching: {
        isLoading: false,
        error: null
      }
    });
    const wrapper = shallow(<Work store={store} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it("renders the loading component when loading", () => {
    store = mockStore({
      pages: {
        work: {}
      },
      fetching: {
        isLoading: true,
        error: null
      }
    });
    const wrapper = shallow(<Work store={store} />);
    const loading = wrapper.find("Loading />");
    expect(loading).toExist;
  });
});
