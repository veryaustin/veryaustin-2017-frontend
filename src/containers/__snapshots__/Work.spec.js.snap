// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Work Component renders the work component 1`] = `
<Work
  fetchPage={[Function]}
  hasErrored={null}
  isLoading={false}
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
  work={Object {}}
/>
`;
